app/services/sp_return_url_resolver.rb

```diff
@@ -0,0 +1,38 @@
+class SpReturnUrlResolver
+  attr_reader :service_provider, :oidc_state, :oidc_redirect_uri
+
+  def initialize(service_provider:, oidc_state:, oidc_redirect_uri:)
+    @service_provider = service_provider
+    @oidc_state = oidc_state
+    @oidc_redirect_uri = oidc_redirect_uri
+  end
+
+  def return_to_sp_url
+    if oidc_redirect_uri.present?
+      oidc_access_denied_redirect_url
+    elsif service_provider.return_to_sp_url.present?
+      service_provider.return_to_sp_url
+    else
+      inferred_redirect_url
+    end
```

Comment:
with the change to make `oidc_access_denied_redirect_url` return nil, we could do:

```suggestion
    oidc_access_denied_redirect_url ||
      service_provider.return_to_sp_url.presence ||
      inferred_redirect_url
```

Reply:
Added in a4f3d073c